File: 4/main.py

def bitshift(section):
    start, finish = section.split('-')
    size = int(finish) - int(start) + 1
    size_mask = 0
    for n in range(size):
        size_mask |= 1 << n + 1
    result = size_mask << (int(start))
    result = result >> 2
    # todo, find out why it's shifted too much
    return result


f = open("4/input.txt")
lines = iter(f.readlines())
result = 0
result2 = 0
for sections in lines:
    elf1, elf2 = sections.split(',')
    elf1_bit = bitshift(elf1)
    elf2_bit = bitshift(elf2)
    or_result = elf1_bit | elf2_bit
    print(sections)
    print(f"{elf1_bit:#0128b}")
    print(f"{elf2_bit:#0128b}")

    if elf1_bit == or_result or elf2_bit == or_result:
        result += 1
    if not elf1_bit & elf2_bit == 0:
        result2 += 1

print(result)
print(result2)


f.close()

File: 4/src/main.rs
use std::env;
use std::fs;


fn bitshift(section: &str) -> u128 {
    // println!("{section}");
    let (start, finish) = section.split_once('-').unwrap();
    let start_bin = start.parse::<u128>().unwrap();
    let finish_bin = finish.parse::<u128>().unwrap();
    let size = finish_bin - start_bin;
    let mask: u128 = 2u128.pow(u32::try_from(size + 1).unwrap()) -1;
    let result = mask << start_bin;
    // println!("{:0>128b}", start_bin);
    // println!("{:0>128b}", finish_bin);
    // println!("{:0>128b}", mask);
    // println!("{:0>128b}", result);
    return result;
}


fn main() {
    let mut current_dir = env::current_dir().unwrap().to_str().unwrap().to_owned();
    let today = "/4";
    if !current_dir.ends_with(today) {
        current_dir += today
    }
    let contents = fs::read_to_string("4/input.txt").expect("couldn't read file");
    let mut result = 0;
    let mut result2 = 0;
    let lines = contents.lines();
    for l in lines.into_iter() {
        // println!("{l}");
        let (elf1, elf2) = l.split_once(',').unwrap();
        let elf1_bit = bitshift(elf1);
        let elf2_bit = bitshift(elf2);
        let or_result = elf1_bit | elf2_bit;
        if (elf1_bit == or_result) || (elf2_bit == or_result) {
            result += 1;
        }
        if (elf1_bit & elf2_bit) != 0u128 {
            result2 += 1;
        }
    }
    println!("{result}");
    println!("{result2}");
}
